accounts/store.go

```diff
@@ -0,0 +1,443 @@
+	var accounts []*OffChainBalanceAccount
+	err := s.db.View(func(tx *bbolt.Tx) error {
+		// This function will be called in the ForEach and receive
+		// the key and value of each account in the DB. The key, which
+		// is also the ID is not used because it is also marshaled into
+		// the value.
+		readFn := func(_, v []byte) error {
+			if v == nil {
+				return nil
+			}
+
+			account, err := deserializeAccount(v)
+			if err != nil {
+				return err
+			}
+
+			accounts = append(accounts, account)
+			return nil
+		}
+
+		// We know the bucket should exist since it's created when
+		// the account storage is initialized.
+		return tx.Bucket(accountBucketName).ForEach(readFn)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return accounts, nil
+}
+
+// RemoveAccount finds an account by its ID and removes it from the DB.
+func (s *Store) RemoveAccount(id AccountID) error {
+	err := s.db.Update(func(tx *bbolt.Tx) error {
+		// First see if account exists.
+		bucket := tx.Bucket(accountBucketName)
+		if bucket == nil {
+			return ErrAccountBucketNotFound
+		}
+
+		account := bucket.Get(id[:])
+		if len(account) == 0 {
+			return ErrAccNotFound
+		}
+
+		// Then remove it, if it exists.
+		return bucket.Delete(id[:])
+	})
+	return err
+}
+
+// checkBalance ensures an account is valid and has a balance equal to
+// or larger than the amount that is required.
+func (s *Store) checkBalance(id AccountID,
+	requiredBalance lnwire.MilliSatoshi) error {
+
+	// Check that the account exists, it hasn't expired and has sufficient
+	// balance.
+	account, err := s.GetAccount(id)
+	if err != nil {
+		return err
+	}
+
+	if account.HasExpired() {
+		return ErrAccExpired
+	}
+
+	if account.CurrentBalance < requiredBalance {
+		return ErrAccBalanceInsufficient
+	}
+
+	return nil
+}
+
+// chargeAccount subtracts the given amount from an account's balance.
+func (s *Store) chargeAccount(id AccountID, paymentHash lntypes.Hash,
+	amount lnwire.MilliSatoshi) error {
+
+	// Check that the account exists and has sufficient balance.
+	account, err := s.GetAccount(id)
+	if err != nil {
+		return err
+	}
+	if err := s.checkBalance(id, amount); err != nil {
+		return err
+	}
+
+	// Update the account and store it in the database.
+	account.CurrentBalance -= amount
+	account.Payments[paymentHash] = struct{}{}
+	return s.storeAccount(account)
+}
+
+// associateInvoice associates a generated invoice with the given account,
+// making it possible for the account to be credited in case the invoice is
+// paid.
+func (s *Store) associateInvoice(id AccountID, hash lntypes.Hash) error {
+	account, err := s.GetAccount(id)
+	if err != nil {
+		return err
+	}
+
+	account.Invoices[hash] = struct{}{}
+	return s.storeAccount(account)
+}
+
+// CreditAccount checks if a paid invoice was associated with any existing
+// account. If an account is found that is still valid, it gets the invoice
+// amount credited.
+func (s *Store) CreditAccount(paidInvoice lntypes.Hash,
```

Comment:
would it perhaps be worth keeping an in-memory map of `map[invoice-hash] accountID` so that this is quicker to check?

Reply:
Yeah, good point. Since we need to listen for all invoices anyway in the service we might as well keep track of them there.